### PR TITLE
[SPARK-42224][CONNECT] Migrate `TypeError` into error framework for Spark Connect functions

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to migrate `TypeError` into error framework for Spark Connect functions.

### Why are the changes needed?

To improve errors by leveraging the PySpark error framework for Spark Connect functions.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Fixed & added UTs.

### CODE REVIEW

File: python/pyspark/sql/tests/test_functions.py [hunk 1]
```diff
@@ -1021,15 +1021,14 @@ def test_assert_true(self):
         with self.assertRaisesRegex((Py4JJavaError, SparkConnectException), "2000000"):
             df.select(assert_true(df.id < 2, df.id * 1e6)).toDF("val").collect()
 
-        with self.assertRaises((PySparkTypeError, TypeError)) as pe:
+        with self.assertRaises(TypeError) as pe:
```
Comment (itholic) [inline python/pyspark/sql/tests/test_functions.py:1024]: This addressed https://github.com/apache/spark/pull/39450/files#r1088566676.

File: python/pyspark/sql/tests/test_functions.py [hunk 2]
```diff
@@ -1042,15 +1041,14 @@ def test_raise_error(self):
         with self.assertRaisesRegex((Py4JJavaError, SparkConnectException), "barfoo"):
             df.select(raise_error("barfoo")).collect()
 
-        with self.assertRaises((PySparkTypeError, TypeError)) as pe:
+        with self.assertRaises(PySparkTypeError) as pe:
```
Comment (itholic) [inline python/pyspark/sql/tests/test_functions.py:1044]: ditto.

File: python/pyspark/sql/tests/test_functions.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/sql/tests/test_functions.py:1024]: FYI: @HyukjinKwon and @techaddict

**Review by techaddict — APPROVED** [review]
@itholic thanks, LGTM!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Pull request closed without merging**